### PR TITLE
Add MathDial

### PR DESCRIPTION
**Pull request by mmhamdy**
Hi,
This PR adds the MathDial (with the downloader, preparer, and mapper). All tests passed successfully.

**Note**: Since the data format expects only the roles of `user` or `assistant`, I've replaced the teacher role (human annotator) with `user`, and the student role (ChatGPT) with `assistant` in the prepare function. But, since this is a tutoring dataset that aims to create automatic dialogue tutors, I think the reverse makes more sense. 

Tell me if you agree and I'll make the required changes.

### CODE REVIEW

File: data_summaries/MathDial.json [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        "Creators": [
+            "ETH Zürich",
+            "Technische Universitat Darmstadt",
+            "Nanyang Technological University"
+        ],
+        "Licenses": [
+            {
+                "License": "CC BY-SA 4.0",
+                "License URL": "https://github.com/eth-nlped/mathdial#example"
+            },
```
Comment (shayne-longpre) [inline data_summaries/MathDial.json:44]: Also add the GSM8K MIT license: https://huggingface.co/datasets/gsm8k#licensing-information

File: src/preparers.py [hunk 2]
```diff
@@ -864,3 +864,17 @@ def prepare_bactrianx(row):
...
     ]
+
+def prepare_mathdial(row):
+    conversation = row["conversation"].split("|EOM|")
+    parent = "mathdial"
+    messages = []
+    for i, turn in enumerate(conversation):
+        colon_index = turn.find(":")
+        messages.append({
+            "from": "user" if turn[:colon_index].strip() == "Teacher" else "assistant",
```
Comment (shayne-longpre) [inline src/preparers.py:875]: If you think it makes more sense to switch user and assistant, so that the user is the student, then let's do that.

**Review by shayne-longpre — APPROVED** [review]
Approved with 2 minor suggestions. Thanks!

File: data_summaries/MathDial.json (hunk 1 above)
Comment (mmhamdy) [reply data_summaries/MathDial.json:44]: Great! Also, I will expand the text sources once the other math-related PR gets merged. I will add most of the sources in Orca Math, so I don't want to duplicate the constants here.

Comment (shayne-longpre) [reply data_summaries/MathDial.json:46]: Should be ready to make this change and merge?

**Comment by mmhamdy** [comment]
Ready to merge.

**Pull request merged**